File: PartyClickerIos/PartyClickerIos/ViewController.m
//
//  ViewController.m
//  PartyClickerIos
//
//  Created by Kevin Galligan on 1/5/17.
//  Copyright © 2017 Kevin Galligan. All rights reserved.
//

#import "ViewController.h"
#import "ComKgalliganPartyclickerDataDatabaseHelper.h"
#import "ComKgalliganPartyclickerAppManager.h"
#import "ComKgalliganPartyclickerDataParty.h"
#import "ComKgalliganPartyclickerDataPartyPresenter.h"
#import "ShowPartyViewController.h"
#import "java/util/List.h"
#import "java/util/ArrayList.h"

@interface ViewController ()
@property (weak, nonatomic) IBOutlet UITableView *tableView;
@property (weak, nonatomic) IBOutlet UITextField *partyNameField;
@property (nonatomic, strong) JavaUtilArrayList *parties;
@end

@implementation ViewController
- (IBAction)createParty:(id)sender {
    ComKgalliganPartyclickerDataDatabaseHelper* dh = [ComKgalliganPartyclickerDataDatabaseHelper getInstanceWithAndroidContentContext:[ComKgalliganPartyclickerAppManager getContext]];
    
    ComKgalliganPartyclickerDataParty* party = [dh createPartyWithNSString:[self.partyNameField text]];
    [self showParty:party];
}

- (void) showParty:(ComKgalliganPartyclickerDataParty*)party
{
    [self refreshParties];
    [self.tableView reloadData];
    [self performSegueWithIdentifier:@"ShowParty" sender:party];
}



- (void) refreshParties
{
    ComKgalliganPartyclickerDataDatabaseHelper* dh = [ComKgalliganPartyclickerDataDatabaseHelper getInstanceWithAndroidContentContext:[ComKgalliganPartyclickerAppManager getContext]];
    self.parties = (JavaUtilArrayList*)[dh allParties];
}

- (void)viewDidLoad {
    [super viewDidLoad];
    [self refreshParties];
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    return [self.parties size];
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    static NSString *simpleTableIdentifier = @"SimpleTableItem";
    
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:simpleTableIdentifier];
    
    if (cell == nil) {
        cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:simpleTableIdentifier];
    }
    
    ComKgalliganPartyclickerDataParty* party = (ComKgalliganPartyclickerDataParty*)[self.parties getWithInt:(jint)indexPath.row];
    
    cell.textLabel.text = party->name_;
    
    return cell;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    [tableView deselectRowAtIndexPath:indexPath animated:NO];
    ComKgalliganPartyclickerDataParty* party = (ComKgalliganPartyclickerDataParty*)[self.parties getWithInt:(jint)indexPath.row];
    [self performSegueWithIdentifier:@"ShowParty" sender:party];
}

- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender
{
    if ([[segue identifier] isEqualToString:@"ShowParty"]) {
        ShowPartyViewController *detailVC = segue.destinationViewController;
        ComKgalliganPartyclickerDataParty *party = (ComKgalliganPartyclickerDataParty *)sender;
        detailVC.presenter = [[ComKgalliganPartyclickerDataPartyPresenter alloc] initWithInt:party->id__];
    }
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}


@end

File: ios/ios/ViewController.swift
//
//  ViewController.swift
//  ios
//
//  Created by Kevin Galligan on 4/30/17.
//  Copyright © 2017 Kevin Galligan. All rights reserved.
//

import UIKit

class ViewController: UIViewController, UITableViewDelegate, UITableViewDataSource, PPPartyListPresenter_UiInterface {
    @IBOutlet weak var tableView: UITableView!
    @IBOutlet weak var partyNameTextView: UITextField!
    @IBOutlet weak var editButton: UIBarButtonItem!
    
    var parties: JavaUtilList!
    var presenter: PPPartyListPresenter!
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        presenter = PPPartyListPresenter()
        PAppManager.getInstance().getDaggerComponent().inject(with: presenter)
        presenter.applyUiInterface(with: self)
        presenter.callRefreshPartyList()
        
        tableView.register(UITableViewCell.classForKeyedArchiver(), forCellReuseIdentifier: "SimpleTableItem")
        styleEditButton()        
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    override func viewWillAppear(_ animated: Bool) {
        presenter.callRefreshPartyList()
    }
    
    @IBAction func editButton(_ sender: Any) {
        tableView.isEditing = !tableView.isEditing
        styleEditButton()
    }
    
    @IBAction func createPartyClicked(_ sender: Any) {
        presenter.createParty(with: partyNameTextView.text)
        partyNameTextView.text = ""
        presenter.callRefreshPartyList()
    }
    
    func styleEditButton(){
        editButton.title = tableView.isEditing ? "Done" : "Edit"
    }

    func numberOfSections(in tableView: UITableView) -> Int {
        return 1
    }
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        if(parties == nil){
            return 0
        }
        else{
            return Int(parties.size())
        }
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: "SimpleTableItem", for: indexPath)

        let party = parties.getWith(jint(indexPath.row)) as! PDParty
        cell.textLabel?.text = party.description()
        
        return cell
    }
    
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        presenter.callParty(with: jint((parties.getWith(jint(indexPath.row)) as! PDParty).getId()))
        tableView.deselectRow(at: indexPath, animated: false)
    }
    
    func refreshPartyList(with partyList: JavaUtilList!) {
        self.parties = partyList
        tableView.reloadData()
        styleEditButton()
    }
    
    func showParty(with party: PDParty!) {
        performSegue(withIdentifier: "ShowParty", sender: party)
    }
    
    func tableView(_ tableView: UITableView, commit editingStyle: UITableViewCellEditingStyle, forRowAt indexPath: IndexPath) {
        if (editingStyle == UITableViewCellEditingStyle.delete) {
            //[self.objects removeObjectAtIndex:indexPath.row];
            //[tableView deleteRowsAtIndexPaths:@[indexPath] withRowAnimation:UITableViewRowAnimationFade];
            presenter.deleteParty(with: (parties.getWith(jint(indexPath.row)) as! PDParty).getId())
            tableView.isEditing = false
        }
        
    }
    
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        if (segue.identifier == "ShowParty") {
            
//            let detailVC = segue.destination as! UINavigationController
            let partyVC = segue.destination as! PartyViewController
            let party = sender as! PDParty
            partyVC.party = party
        }
    }
}

